render_hawk_2/renderer_class.cs

```diff
@@ -0,0 +1,466 @@
+        int i = 0;
+
+        if (background_worker.CancellationPending)
+        {
+            return output_image;
+        }
+
+        for (int y = 0; y < scene.image_height; y++)
+        {
+            for (int x = 0; x < scene.image_width; x++)
+            {
+                pixel_color = renderer_class.sample_pixel (scene,
+                                                           x,
+                                                           y,
+                                                           random,
+                                                           background_worker);
+                
+                for (int n = 0; n < 3; n++)
+                {
+                    output_image[i + n] = pixel_color[2 - n];
+                }
+
+                i += 3;
+            }
+
+            i += scene.image_width % 4;
+        }
+
+        return output_image;
+    }
+
+    private static double[] sample_pixel (scene_class scene, int x, int y, System.Random random, System.ComponentModel.BackgroundWorker background_worker)
+    {
+        double[] output_pixel = new double[3];
+
+        double[] position = new double[3];
+        double[] vector = new double[3];
+
+        position = scene.camera.position;
+        vector = scene.camera.get_vector (scene, x, y);
+
+        output_pixel = get_total_radiance (scene, x, y, random, position, vector);
+
+        return output_pixel;
+
+        /*
+        double[][] output_pixel = new double[5][];
+        double[] teh_uber_output_pixel = new double[3];
+        for (int i = 0; i < output_pixel.GetLength (0); i++)
+        {
+            output_pixel[i] = new double[3];
+        }
+
+
+        double[] position = new double[3];
+        double[] vector = new double[3];
+
+        for (int i = 0; i < output_pixel.GetLength (0); i++)
+        {
+            position = scene.camera.position;
+            vector = scene.camera.get_vector (scene, x, y);
+
+            output_pixel[i] = get_total_radiance (scene, x, y, random, position, vector);
+        }
+
+        teh_uber_output_pixel = output_pixel[0];
+        for (int i = 1; i < output_pixel.GetLength (0); i++)
+        {
+            teh_uber_output_pixel = vector_class.vector_average (teh_uber_output_pixel, output_pixel[i], 1 / ((double)i + 1));
+        }
+
+        return teh_uber_output_pixel;
+        */
+        
+    }
+
+    private static double[] get_total_radiance (scene_class scene, int x, int y, System.Random random, double[] position, double[] vector)
+    {
+        double[] total_color_radiance = new double[3];
+        double[] vector_t = new double[3];
+        double[] vector_s = new double[3];
+        double[] multiplier = new double[3];
+        for (int i = 0; i < 3; i++)
+        {
+            multiplier[i] = 1;
+        }
+        double medium_refractive_index = 1.0008;
+        int triangle_hit = -1;
+        int sphere_hit = -1;
+
+        for (int i = 0; i < scene.ray_depth; i++)

```

Comment:
This would make more sense as a recursive method call which decrements the remaining ray depth.